src/lib/query.ts

```diff
@@ -0,0 +1,38 @@
+import { parseQuery, Query, QueryCodec } from '@pomle/paths';
+
+type ParamCache = { source: string; value: unknown };
+
+export function createParser<T extends QueryCodec>(query: Query<T>) {
+  const memo = new Map<string, ParamCache[]>();
+
+  function getCache(key: string) {
+    if (!memo.has(key)) {
+      memo.set(key, []);
+    }
+    return memo.get(key) as ParamCache[];
```

Comment:
to void casting you can do

function getCache(key: string) {
  let result = memo.get(key);
  if (!result) {
    result = [];
    memo.set(key, result)
  }
  return result;
}